hudi-common/src/main/java/org/apache/hudi/io/storage/HoodieHFileReader.java

```diff
@@ -209,12 +212,23 @@ public R next() {
 
   @Override
   public Option getRecordByKey(String key, Schema readerSchema) throws IOException {
-    HFileScanner scanner = reader.getScanner(false, true);
+    byte[] value = null;
     KeyValue kv = new KeyValue(key.getBytes(), null, null, null);
-    if (scanner.seekTo(kv) == 0) {
-      Cell c = scanner.getKeyValue();
-      byte[] keyBytes = Arrays.copyOfRange(c.getRowArray(), c.getRowOffset(), c.getRowOffset() + c.getRowLength());
-      R record = getRecordFromCell(c, getSchema(), readerSchema);
+
+    synchronized (this) {
+      if (keyScanner == null) {
+        keyScanner = reader.getScanner(true, true);
```

Comment:
is it an overhead to initialize a scanner. if not, why not we initialize in constructor only?

Reply:
Not initializing in constructor because the HoodieHFileReader can be used for other purposes like getAllRecords() or readBloomFilter() or getTotalRecords() (through other code paths) which do not require the KeyScanner to be initialized.

Reply:
gotcha